### PR TITLE
Project Recovery: Sanitise the # charecter in workspace names

### LINKED ISSUE #23248: Project Recovery: Sanitise the # charecter
Sanitize the # character in project recovery.

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
Stops the project recovery from using # in any of it's scripts and also sorts an issue when created workspaces do not match original name list.

**To test:**

* Run the following script:
```
CreateWorkspace(DataX=range(12), DataY=range(12), DataE=range(12), NSpec=4, OutputWorkspace='0Reb#inned')
for i in range(100):
    RenameWorkspace(InputWorkspace='%sReb#inned'%str(i), OutputWorkspace='%sReb#inned'%str(i+1))
for i in range(300):
    CloneWorkspace(InputWorkspace='100Reb#inned', OutputWorkspace='%sCl#o#ne'%str(i))
```
*  Allow project recovery to start and finish
* Crash Mantid using the segfault algorithm
* Restart Mantid; select `Yes` when presented with recovery options
* Check that the workspaces are re-loaded, e.g. see if `232Cl#o#ne` exists as a workspace

Fixes #23248  

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/src/ScriptBuilder.cpp [hunk 1]
```diff
@@ -265,6 +267,9 @@ const std::string ScriptBuilder::buildPropertyString(
 
       prop = propHistory.name() + opener + propHistory.value() + "'";
     }
+    // Sanitise for a comment #
```
Comment (DanNixon) [inline Framework/API/src/ScriptBuilder.cpp]: Why is this needed? '#' will not open a comment if it is enclosed in quotation marks.

Comment (keeeto) [reply Framework/API/src/ScriptBuilder.cpp]: It is to do with the way that the recovery script builder constructs the ordered history. It divides each line of workspace history between the command and the time/date stamp and uses `#` as the separating character. 

Currently, a `# ` in the name causes the script builder to separate at the wrong location.

Comment (DanNixon) [reply Framework/API/src/ScriptBuilder.cpp]: Yes, I actually tried it and found the error.

I think we should fix the hacky Python parsing in `OrderWorkspaceHistory` (this is where the error came from for me and it's nature is as you describe, there may be more instances I don't know about) however as `#` is perfectly valid in a Python string.

There is the [`tokenize`](https://docs.python.org/3.5/library/tokenize.html) library which will correctly read lines of Python and separate out comments.

Comment (keeeto) [reply Framework/API/src/ScriptBuilder.cpp]: @Pasarus - I think that Dan has a point here. Could you try the `tokenize` alternative to this fix, without the changes made here? I think that fix should also reduce the amount of alteration of the original code required.

Comment (Pasarus) [reply Framework/API/src/ScriptBuilder.cpp]: The original code editing was with regards to the way Qt confirmed that all workspaces had been loaded properly, and I simply checked it, to see if a # had been used in the original workspace names and if it had then I would let it continue and through a debug warning that, that had happened, then I would check if it hadn't needed sanitizing normally.

I'm not great with python but I can give it a go.

Comment (Pasarus) [reply Framework/API/src/ScriptBuilder.cpp]: Having re-evaluated I believe it is completely required, in the OrderWorkspaceHistory.py to split based on the # for comments as it removed the date/time from the command, and the idea is that it's stored together. Changing this would require quite a bit of extra work, to restructure how it saves and loads the data.

Comment (DanNixon) [reply Framework/API/src/ScriptBuilder.cpp]: https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py#L87 is exactly the code I have a problem with, it makes a false assumption about how comments work in Python.

If you use `tokenize` to do this parsing you can robustly extract comments from Python source so having a `#` in a string wouldn't be an issue in the first place.

Comment (keeeto) [reply Framework/API/src/ScriptBuilder.cpp]: @DanNixon, looking at `tokenize` - it is a bit more involved than I had thought. Do you have any experience using it? I could save us a lot of time by dropping over to you if you do.

Comment (DanNixon) [reply Framework/API/src/ScriptBuilder.cpp]: @keeeto Yes, I can take a look this morning. Do you know of anywhere else this assumption is made apart from `OrderWorkspaceHistory`?

**Review by keeeto — APPROVED** [review]
I have checked this against the project recovery unscripted tests. However, as I have contributed some of the code, it would be better if we could get another review on this.

**Comment by martyngigg** [comment]
@DanNixon There are some flake8 warnings to fix.

**Comment by keeeto** [comment]
@martyngigg @DanNixon I've looked at the flake 8 warnings associated with this code and fixed them. There are some other warnings coming from code that has not been touched in this PR - `DPDFreduction.py`, `LRReflectivityOutput.py` and `LiquidsReflectometryReduction.py`. I'm not sure what's going on there.

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py [hunk 2]
```diff
@@ -67,38 +68,45 @@ def write_ordered_workspace_history(self):
...
+                        have_comment = any(x[0] == tokenize.COMMENT for x in line)
+                        if have_comment:
+                            all_lines.append(line)
+                        line = []
+                    # Everything in between we care about
+                    elif line is not None:
+                        line.append(t)
+
+        # l[-1][1][1:] is the comment string, with the preceeding hash stripped off
+        all_commands = [(''.join(t[1] for t in l[:-1]), l[-1][1][1:]) for l in all_lines]
```
Comment (Pasarus) [inline Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: @keeeto You can probably replace L94-L101 with
```python
        # l[-1][1][1:] is the comment string, with the preceeding hash stripped off
        unique_commands = [(''.join(t[1] for t in l[:-1]), l[-1][1][1:])
                      for l in all_lines]

        # Remove duplicate commands by casting commands as a set and back to a list
        unique_commands = list(set(unique_commands))

        # Sort the new list on datetime integer
        unique_commands.sort(key=lambda time: (time[1]))
```

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py [hunk 3]
```diff
@@ -67,38 +68,45 @@ def write_ordered_workspace_history(self):
...
-                               for i in all_unique_commands]
+        all_lines = []
+
+        for fn in historyfiles:
+            with open(os.path.join(source_folder, fn)) as f:
+                tokens = tokenize.generate_tokens(f.readline)
+
+                line = None
+                for t in tokens:
+                # Start a new line when we see a name
```
Comment (DanNixon) [inline Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: Indentation is off on this comment.

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py [hunk 4]
```diff
@@ -67,38 +68,45 @@ def write_ordered_workspace_history(self):
...
-
-        # Add lists of histories together
-        all_unique_commands = [command.strip().split('#') for command in commands]
-
-        # Convert the datetime into a sortable integer
-        all_unique_commands = [(i[0], self._concatenate_iso_datetime(i[1]))
-                               for i in all_unique_commands]
+        all_lines = []
+
+        for fn in historyfiles:
```
Comment (DanNixon) [inline Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: The way this list is generated can be replaced by a single call to [`glob.iglob`](https://docs.python.org/2/library/glob.html#glob.iglob).

e.g.
```python
for fn in glob.iglob(os.path.join(self.getPropertyValue(_recovery_folder), '*.py')):
    with open(fn) as f:
        # do stuff here
```

**Review by DanNixon — CHANGES_REQUESTED** [review]
A couple more ways this algorithm can be tidied that may as well be done now.

Also I'd vote for renaming the `write_ordered_workspace_history()` function as writing is a minority of what it actually does. Id go as far to say it may as well just be the `PyExec()` function since all that contains is a call to this function.

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py [hunk 5]
```diff
@@ -37,70 +39,45 @@ def validateInputs(self):
...
-        if not date_time:
-            raise ValueError("No date time stamp was found in the recovery history")
-        self.log().debug("Found datetime stamp: {0}".format(date_time))
-
-        # Remove whitespace and any trailing zeros
-        stipped_time = date_time.strip().rstrip('0')
-        return datetime.datetime.strptime(stipped_time, "%Y-%m-%dT%H:%M:%S.%f")
-
-    def write_ordered_workspace_history(self):
         self.log().debug("Started saving workspace histories")
```
Comment (DanNixon) [inline Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: Given that the start/end of algorithms are logged do we need this?

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py [hunk 6]
```diff
@@ -37,70 +39,45 @@ def validateInputs(self):
...
-
-        # Convert the datetime into a sortable integer
-        all_unique_commands = [(i[0], self._concatenate_iso_datetime(i[1]))
-                               for i in all_unique_commands]
-
-        # Sort the new list on datetime integer
-        all_unique_commands.sort(key=lambda time: (time[1]))
+        all_lines = []
+
+        #for fn in historyfiles:
```
Comment (DanNixon) [inline Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: Old commented code?

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py [hunk 7]
```diff
@@ -37,70 +39,45 @@ def validateInputs(self):
...
+                        have_comment = any(x[0] == tokenize.COMMENT for x in line)
+                        if have_comment:
+                            all_lines.append(line)
+                        line = []
+                    # Everything in between we care about
+                    elif line is not None:
+                        line.append(t)
+
+        # l[-1][1][1:] is the comment string, with the preceeding hash stripped off
+        all_commands = [(''.join(t[1] for t in l[:-1]), l[-1][1][1:]) for l in all_lines]
```
Comment (DanNixon) [inline Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: I think the first element of the tuple needs explaining too, without being familiar with `tokenize` you probably wouldn't realise that this is reconstructing the line minus the comment (last token).

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py [hunk 8]
```diff
@@ -37,70 +39,45 @@ def validateInputs(self):
...
+
+        unique_commands.sort(key=lambda time: (time[1]))
 
         destination = self.getPropertyValue(_destination_file)
 
         self.log().debug("Writing commands to file")
         # Write to file
         with open(destination, 'w') as outfile:
-            for x in all_unique_commands:
+            for x in unique_commands:
```
Comment (DanNixon) [inline Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: Is this just writing the lines back out as they were?

In that case can't we just skip rebuilding the line from the tokens above and just output the full line from the comment token?

**Review by DanNixon — CHANGES_REQUESTED** [review]
Just a few more comments.

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py (hunk 8 above)
Comment (keeeto) [reply Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: @DanNixon I'm not sure I follow. The list `unique_commands` has been stripped of replication and sorted on the timestamp. I can't see how we would get the correct order from the comment token.

Comment (DanNixon) [reply Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: `''.join(t[1] for t in l[:-1])` above is recreating the line originally read from the file minus the comment.

`outfile.write('{} # {} \n'.format(x[0], x[1]))` is formatting a string containing the reconstructed string previously mentioned, a `#` and the comment minus the `#`. The result of this is the same as the full source line that is also stored in the comment token (save the newline).

So really we only need to store the comment tokens, sort them on the comment text and output the full source line (last element of the token tuple) here.

Comment (Pasarus) [reply Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: I believe the idea behind it, is that we are removing duplicate commands from the list so that project recovery doesn't run into issues further down the road. With the method you are suggesting it does not seem to take this into account @DanNixon

Comment (keeeto) [reply Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: No problem - hard to describe these concepts in this forum. If you can drop by tomorrow that'd be great. I'll be out from about 9:25 - 10:10. Otherwise I'm about.

**Comment by keeeto** [comment]
@DanNixon I've re-written to take account of your previous comment. The change is not *exactly* what you suggested, but I think it achieves the same thing. 

During the reading/tokenize step I take the relevant information from the comment token (i.e. command and comment) and store as a tuple of tuples.

This is then cast as a set to remove duplicates and then sorted on the time (comments).

The first member of the tuples (the command) is then written to the history file.

I basically wanted to get away from the comment token structure asap - as I find it hard to follow. I don't think that this should make the solution any less efficient and also negates any (minor) concerns that we had about the token object in Python 3.

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py [hunk 9]
```diff
@@ -37,70 +38,39 @@ def validateInputs(self):
...
         destination = self.getPropertyValue(_destination_file)
 
-        self.log().debug("Writing commands to file")
-        # Write to file
         with open(destination, 'w') as outfile:
-            for x in all_unique_commands:
-                outfile.write('{} # {} \n'.format(x[0], x[1]))
-
+            for x in unique_lines:
+                outfile.write('{}'.format(x[0]))
```
Comment (DanNixon) [inline Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: The string formatting is no longer needed here.

i.e. just `outfile.write(x[0])` will do.

File: Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py [hunk 10]
```diff
@@ -37,70 +38,39 @@ def validateInputs(self):
...
+                    # Start a new line when we see a name
+                    if line is None and t[0] == tokenize.NAME:
+                        line = [t]
+                    # End the line when we see a logical line ending
+                    elif t[0] == tokenize.NEWLINE:
+                        # Only care about the line if it has a comment
+                        have_comment = any(x[0] == tokenize.COMMENT for x in line)
+                        if have_comment:
+                            # line[-1][1][1:] is the comment string, with the preceeding hash stripped off; line[0][4] is the command and the comment
+                            all_lines.append((line[0][4], line[-1][1][1:]))
```
Comment (DanNixon) [inline Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py]: Can we take the command and comment (i.e. first element of the tuple) from the comment token (`line[-1][4]`).

The fact that it is taken from the first token implies a reason it needs to be, when in fact it can be any token from the same line.

**Review by DanNixon — CHANGES_REQUESTED** [review]
Last two comments.

**Comment by DanNixon** [comment]
Sorry, one very last thing. The release notes should go in this PR (I'm not too sure why they were separated into a different PR in the first place).

**Comment by keeeto** [comment]
That's a fair point @DanNixon - I think that @Pasarus maybe wants to merge the other PR into this one, so the release notes will cover both. I'll check the rationale tomorrow and we can decide what to do. I think it would probably be easier to separate the two completely, but it is worth checking.

**Review by DanNixon — APPROVED** [review]
